src/solvers/refinement/string_refinement_util.cpp

```diff
@@ -0,0 +1,414 @@
+/*******************************************************************\
+
+ Module: String solver
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <algorithm>
+#include <numeric>
+#include <functional>
+#include <iostream>
+#include <util/arith_tools.h>
+#include <util/ssa_expr.h>
+#include <util/std_expr.h>
+#include <util/expr_iterator.h>
+#include <util/graph.h>
+#include "string_refinement_util.h"
+
+bool is_char_type(const typet &type)
+{
+  return type.id() == ID_unsignedbv &&
+         to_unsignedbv_type(type).get_width() <= 16;
```

Comment:
Is this constant language-specific?

Reply:
The constant is documented in the header. For now we consider the solver only works for strings whose characters are at most 16 bits so that's fine for java and C. Ideally it should be a parameter of the solver or activated for all primitive types.

Reply:
Use `util/magic.h`. Documentation cannot be read by machines and doing `grep` for arbitrary terms is hard.